### PR TITLE
[KOGITO-5449] Fix JandexProtoGenerator StackOverflow and ProtoGenerator refactor

### PR DESCRIPTION
**Pull request by danielezonca**
https://issues.redhat.com/browse/KOGITO-5449

This PR fix a StackOverflowException in case of recursive types and also includes multiple refactoring/improvements to ProtoGenerator in general
- Create a common test suite for both `JandexProtoGenerator` and `ReflectionProtoGenerator`
- Fix enum different behaviour introduced by https://github.com/kiegroup/kogito-runtimes/pull/654 
- Fix different behaviour introduced by https://github.com/kiegroup/kogito-runtimes/pull/1468
- Fix bug in package handling: top level types were using proto package instead of their own real packages
- Unified proto generation to have a single common path: all public methods reach `internalGenerate` that is the only entry point
- Moved to `AbstractProtoGenerator` some common logic (to be honest ~all the code in `ReflectionProtoGenerator` and `JandexProtoGenerator` is a duplication from a functional perspective but going deeper with common methods requires to "replace" all interaction with `Class<?>` and `ClassInfo` to common abstraction)

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [ ] You have read the [contributors guide](CONTRIBUTING.md)
- [ ] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config)
- [ ] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [ ] Pull Request title contains the target branch if not targeting main: `[0.9.x] KOGITO-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Run all builds</b>  
  Please add comment: <b>Jenkins retest this</b>

* <b>Run (or rerun) specific test(s)</b>  
  Please add comment: <b>Jenkins (re)run [runtimes|optaplanner|apps|examples] tests</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Run (or rerun) LTS specific test(s)</b>  
  Please add comment: <b>Jenkins (re)run [runtimes|optaplanner|apps|examples] LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

* <b>Run (or rerun) native specific test(s)</b>  
  Please add comment: <b>Jenkins (re)run [runtimes|optaplanner|apps|examples] native</b>

* <b>Full Kogito testing</b> (with cloud images and operator BDD testing)  
  Please add comment: <b>Jenkins run BDD</b>  
  <b>This check should be used only if a big change is done as it takes time to run, need resources and one full BDD tests check can be done at a time ...</b>
</details>

### CODE REVIEW

File: kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/persistence/proto/ProtoComponent.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kie.kogito.codegen.process.persistence.proto;
+
+public abstract class ProtoComponent {
```
Comment (danielezonca) [inline kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/persistence/proto/ProtoComponent.java:18]: Created a common class with common fields between `ProtoEnum` and `ProtoMessage `

File: quarkus/extensions/kogito-quarkus-processes-extension/kogito-quarkus-processes-deployment/src/main/java/org/kie/kogito/quarkus/processes/deployment/JandexProtoGenerator.java [hunk 2]
```diff
@@ -188,27 +174,32 @@ protected ProtoMessage messageFromClass(Proto proto, ClassInfo clazz, IndexView
...
-            name = altName;
+    @Override
+    protected ProtoEnum enumFromClass(Proto proto, ClassInfo clazz) {
+        try {
+            return extractName(clazz)
+                    .map(name -> {
+                        ProtoEnum modelEnum = new ProtoEnum(name, clazz.name().prefix().toString());
+                        clazz.fields().stream()
+                                .filter(f -> !f.name().startsWith("$"))
+                                .sorted(Comparator.comparing(FieldInfo::name))
```
Comment (danielezonca) [inline quarkus/extensions/kogito-quarkus-processes-extension/kogito-quarkus-processes-deployment/src/main/java/org/kie/kogito/quarkus/processes/deployment/JandexProtoGenerator.java:185]: This comparator always sort the values of an enum and prevent different behaviour between Jandex and Reflection (introduced by https://github.com/kiegroup/kogito-runtimes/pull/654)

**Review by cristianonicolai — APPROVED** [review]

**Comment by danielezonca** [comment]
jenkins rerun apps
jenkins rerun examples

**Comment by danielezonca** [comment]
Jenkins rerun apps tests
Jenkins rerun examples tests

**Review by ricardozanini — APPROVED** [review]
Thanks for this refactoring. I can't see anything relevant in the code that needs to be changed for this PR. The new approach seems fine. However, I believe we need to coordinate what to do for the mid-term, mainly because the BPMN stateful process should be our target now.

**Review by tiagodolphine — APPROVED** [review]
LGTM, nice refactoring.
Thanks @danielezonca

**Comment by danielezonca** [comment]
Jenkins rerun examples tests

**Comment by ricardozanini** [comment]
Jenkins rerun examples tests

**Comment by danielezonca** [comment]
Jenkins rerun examples tests

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Pull request merged**